File: 0144-binary-tree-preorder-traversal/0144-binary-tree-preorder-traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
// iterative -> O(n)
    vector<int> preorderTraversal(TreeNode* root) {
        vector<int> ans;
        if(root == NULL)return {};

        stack<TreeNode*> s;
        s.push(root);
        
        while(!s.empty()){
            TreeNode* top = s.top();
            s.pop();
            ans.push_back(top->val);

            TreeNode* left = top->left;
            TreeNode* right = top->right;
            
            if(right)s.push(right);
            if(left)s.push(left);
        }
       
        return ans;
    }
};

File: 0144-binary-tree-preorder-traversal/0144-binary-tree-preorder-traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    List<Integer> preOrder = new ArrayList<>();
    public void dfs(TreeNode root){
        if(root == null)return;
        
        preOrder.add(root.val);
        dfs(root.left);
        dfs(root.right);
    }
    public List<Integer> preorderTraversal(TreeNode root) {
        // dfs(root);
        if(root == null)return preOrder;

        Stack<TreeNode> s = new Stack<>();
        s.push(root);

        while(!s.isEmpty()){
            TreeNode peek = s.pop();

            if(peek != null){
                preOrder.add(peek.val);
                s.push(peek.right);
                s.push(peek.left);
            }

        }
        return preOrder;
    }
}